text/0000-generic-integers.md

```diff
@@ -0,0 +1,437 @@
+
+## Uncommonly sized integers
+
+One other side effect of having `u<N>` and `i<N>` is that we can represent a lot more types than before. For example, `u<7>` is just a seven-bit integer, which we might use to represent an ASCII character. That said, using fewer bits doesn't necessarily mean you'll use up fewer bits of memory-- for example, `u<7>` by itself still requires a `u<8>` in memory.
+
+Overall, you should expect integers where `N` is not a power of two to take up more size/alignment than their bits might imply. There may be ways of packing the bits together in a way that optimizes the amount of space used, but once you read those values into a `u<N>` or `i<N>`, this is no longer the case.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Basic semantics
+
+The compiler will gain the built-in integer types `u<N>` and `i<N>`, where `const N: usize`. These be identical to existing `uN` and `iN` types wherever possible, e.g. `u<8> == u8`.
+
+`usize`/`isize` continue to be separate from `u<N>`/`i<N>`, where `N` is the target pointer width, as this also applied to `uN` and `iN`. `bool` will also be made different from `u<1>` for similar reasons, even though casting `false` to `0` and `true` to `1` will remain possible.
+
+`u<N>` are able to store integers in the range `0..2.pow(N)` and `i<N>` are able to store integers in the range `-2.pow(N-1)..2.pow(N-1)`. The cheeky specificity of "integers in the range" ensures that, for `i<0>`, the range `-0.5..0.5` only contains the integer zero; in general, `u<0>` and `i<0>` will need to be special-cased anyway, as they must be ZSTs.
+
+It's always valid to `as`-cast `u<N>` or `i<N>` to `u<M>` or `i<M>`, and the usual sign extension or truncation will occur depending on the bits involved. A few additional casts which are possible:
+
+* from `bool` to `u<N>` or `i<N>`
+* from `char` to `u<N>` or `i<N>`
+* from `u<1>` to `bool`
+* from `u<N>` to `char`, where `N < 16`
+
+Note that casting directly to `u<0>` or `i<0>` is still allowed, to avoid forcing users to special-case their code. See later notes on possible lints.
+
+Integer literals can be automatically coerced to `u<N>` and `i<N>`, although generic `iN` and `uN` suffixes are left out for future RFCs. When coercing literals to an explicit `u<N>` or `i<N>`, the `overflowing_literals` lint should trigger as usual, although this should not apply for generic code. See later notes on possible lint changes.
+
+In general, operations on `u<N>` and `i<N>` should work the same as they do for existing integer types, although the compiler may need to special-case `N = 0` and `N = 1` if they're not supported by the backend.
+
+When stored, `u<N>` should always zero-extend to the size of the type and `i<N>` should always sign-extend. This means that any padding bits for `u<N>` can be expected to be zero, but padding bits for `i<N>` may be either all-zero or all-one depending on the sign.
```

Comment:
Please clarify this to say what exactly happens when I transmute e.g. `255u8` to `u<7>` (and similar to `i<N>`). I assume it is UB, i.e., the validity invariant of these types says that the remaining bits are zero-extended / sign-extended, but the RFC should make that explicit.

Reply:
Yeah, I'm not quite sure what a better name is; it's the same as `rustc_layout_scalar_valid_range`, which is UB if the bits are invalid.

Reply:
I guess that since this is the reference description, calling them niche bits would be more appropriate? Would that feel reasonable?

Reply:
No. Niche bits are an implementation detail of the enum layout algorithm, and mostly not documented.

Just describe what the valid representations of of values of this type are, i.e., what should go into [this section](https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html#invalid-values) about these types.